python/ml4ir/base/features/feature_fns/categorical.py

```diff
@@ -215,6 +220,105 @@ def categorical_embedding_with_vocabulary_file(feature_tensor, feature_info, fil
     return embedding
 
 
+class CategoricalDropout(layers.Layer):
+    """Custom Dropout class for categorical indices"""
+
+    def __init__(self, dropout_rate, seed=None, **kwargs):
+        """
+        Args:
+            dropout_rate: fraction of units to drop, i.e., set to OOV token 0
+            seed: random seed for sampling
+        """
+        super(CategoricalDropout, self).__init__(**kwargs)
+        self.dropout_rate = dropout_rate
+        self.seed = seed
+
+    def get_config(self):
+        config = super(CategoricalDropout, self).get_config()
+        config.update({"dropout_rate": self.dropout_rate, "seed": self.seed})
+        return config
+
+    def call(self, inputs, training=None):
+        """
+        At training time, mask indices to 0 at dropout_rate
+
+        This works similar to tf.keras.layers.Dropout without the scaling
+        Ref: https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dropout
+
+        Example:
+        inputs: [[1, 2, 3], [4, 1, 2]]
+        dropout_rate = 0.5
+
+        When training, output: [[0, 0, 3], [0, 1, 2]]
+        When testing, output: [[1, 2, 3], [4, 1, 2]]
```

Comment:
👍 👍

Reply:
Could we add a unit-test exactly for this?
with dropout_rate=1, 0.5 and 0?

Reply:
I can't test dropout because it doesn't promise 50% will be dropped. It is random. So the only way I could test it was set the rate to 0.999